src/biotite/structure/io/pdbx/convert.py

```diff
@@ -677,6 +677,59 @@ def _find_matches(query_arrays, reference_arrays):
     return match_indices
 
 
+def _find_matches_by_dict(query_arrays, reference_arrays):
+    """
+    For each index in the `query_arrays` find the indices in the
+    `reference_arrays` where all query values match the reference counterpart.
+    If no match is found for a query, the corresponding index is -1.
+    """
+    # Convert reference arrays to a dictionary for O(1) lookups
+    reference_dict = {}
+    unambiguously_keys = set()
+    for idx, col in enumerate(np.stack(reference_arrays, axis=-1)):
```

Comment:
Running `np.stack()` forces the reference_arrays to the same dtype, probably still requiring time consuming conversion to string and losing the type information. For the purpose required here, `zip()` should achieve the same, right?

Furthermore, you are iterating over the *rows* of the `atom_site` (reference) and `struct_conn` (query) categories instead of *columns*, if I am not mistaken.

```suggestion
    for idx, row in enumerate(zip(*reference_arrays)):
```